### PR TITLE
MPI changes fort trilinos and containers

### PR DESCRIPTION
**Pull request by roigcarlo**
**📝 Description**
Extending #9644:
- Trilinos become an optional component (unless compiling trilinos)
- Created a docker to generate wheels using ABI compatible mpi for linux (through mpich, no 100% guarantees it will work on all systems but should be pretty reliable)

Since the detection of mpi is done through the existance of the mpi folder in the module, I've just changes the way the wheelbuilder works and by default it tries to copy everything installed in the KratosMultiphysics folder in the install dir.

This technically overrides the "included_modules" section in the core configuration json. Once I find a better solution I will change it.

**🆕 Changelog**
- Making Trilinos a optional component in MPI compilations
- Created Docker builder for MPI-ready wheels for linux.

### CODE REVIEW

File: scripts/wheels/linux/configure_mpi.sh [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+add_app ${KRATOS_APP_DIR}/DEMApplication
+add_app ${KRATOS_APP_DIR}/ContactStructuralMechanicsApplication
+add_app ${KRATOS_APP_DIR}/ParticleMechanicsApplication;
+add_app ${KRATOS_APP_DIR}/ConvectionDiffusionApplication;
+add_app ${KRATOS_APP_DIR}/DamApplication;
+add_app ${KRATOS_APP_DIR}/PoromechanicsApplication;
+add_app ${KRATOS_APP_DIR}/FSIApplication;
+add_app ${KRATOS_APP_DIR}/SwimmingDEMApplication;
+# add_app ${KRATOS_APP_DIR}/ExternalSolversApplication;
+add_app ${KRATOS_APP_DIR}/EigenSolversApplication;
```
Comment (philbucher) [inline scripts/wheels/linux/configure_mpi.sh]:
```suggestion
```

those are super outdated

File: scripts/wheels/linux/configure_mpi.sh [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+-DCMAKE_INSTALL_PREFIX=$2                                              \
+-DUSE_TRIANGLE_NONFREE_TPL=ON                                          \
+-DUSE_MPI=ON                                                           \
+-DCMAKE_C_COMPILER=/usr/lib64/mpich/bin/mpicc                          \
+-DCMAKE_CXX_COMPILER=/usr/lib64/mpich/bin/mpicxx                       \
+-DCMAKE_CXX_FLAGS="-msse3 -std=c++11 "                                 \
+-DCMAKE_C_FLAGS="-msse3"                                               \
+-DBOOST_ROOT="/workspace/boost/boost_1_71_0"                           \
+-DLAPACK_LIBRARIES="/usr/lib64/liblapack.so.3"                         \
+-DBLAS_LIBRARIES="/usr/lib64/libblas.so.3"                             \
```
Comment (philbucher) [inline scripts/wheels/linux/configure_mpi.sh]:
```suggestion
```
unused

File: scripts/wheels/linux/configure_mpi.sh [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+# Clean
+rm -rf "${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}/cmake_install.cmake"
+rm -rf "${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}/CMakeCache.txt"
+rm -rf "${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}/CMakeFiles"
+
+${CMAKE} -H"${KRATOS_SOURCE}" -B"${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}" \
+-DCMAKE_INSTALL_PREFIX=$2                                              \
+-DUSE_TRIANGLE_NONFREE_TPL=ON                                          \
+-DUSE_MPI=ON                                                           \
+-DCMAKE_C_COMPILER=/usr/lib64/mpich/bin/mpicc                          \
```
Comment (philbucher) [inline scripts/wheels/linux/configure_mpi.sh:61]: why mpich and not openmpi?

**Review by philbucher — COMMENTED** [review]
cool!
Minor comments afterward I will approve

File: kratos/KratosMultiphysics.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {
 	"wheel_name": "KratosMultiphysics",
-	"included_modules": ["response_functions"],
-	"included_binaries": ["Kratos.*", "KratosCore.*", "KratosCore.*", "zlib.*", "libz.*"],
+    "included_modules":  ["response_functions"],
+	"included_binaries": ["Kratos.*", "KratosCore.*", "KratosCore.*", "KratosMPICore.*", "KratosMPI.*", "zlib.*", "libz.*"],
```
Comment (philbucher) [inline kratos/KratosMultiphysics.json:4]: this also works for a serial compilation?

File: scripts/wheels/linux/configure_mpi.sh (hunk 2 above)
Comment (roigcarlo) [reply scripts/wheels/linux/configure_mpi.sh]: mm you are right!

File: kratos/KratosMultiphysics.json (hunk 4 above)
Comment (roigcarlo) [reply kratos/KratosMultiphysics.json:4]: Yep, this is a filter list. It looks for the files in the folder matching any of those. If you don't compile it won't be copied.

**Pull request merged**